Sources/GRPC/LengthPrefixedMessageWriter.swift

```diff
@@ -86,99 +86,42 @@ internal struct LengthPrefixedMessageWriter {
     buffer: ByteBuffer,
     allocator: ByteBufferAllocator,
     compressed: Bool = true
-  ) throws -> ByteBuffer {
+  ) throws -> (ByteBuffer, ByteBuffer?) {
     if compressed, let compressor = self.compressor {
-      return try self.compress(buffer: buffer, using: compressor, allocator: allocator)
-    } else if buffer.readerIndex >= 5 {
-      // We're not compressing and we have enough bytes before the reader index that we can write
-      // over with the compression byte and length.
-      var buffer = buffer
-
-      // Get the size of the message.
-      let messageSize = buffer.readableBytes
-
-      // Move the reader index back 5 bytes. This is okay: we validated the `readerIndex` above.
-      buffer.moveReaderIndex(to: buffer.readerIndex - 5)
-
-      // Fill in the compression byte and message length.
-      buffer.setInteger(UInt8(0), at: buffer.readerIndex)
-      buffer.setInteger(UInt32(messageSize), at: buffer.readerIndex + 1)
-
-      // The message bytes are already in place, we're done.
-      return buffer
+      let compressedAndFramedPayload = try self.compress(
+        buffer: buffer,
+        using: compressor,
+        allocator: allocator
+      )
+      return (compressedAndFramedPayload, nil)
+    } else if buffer.readableBytes > Self.singleBufferSizeLimit {
+      // Buffer is larger than the limit for emitting a single buffer: create a second buffer
+      // containing just the message header.
+      var prefixed = allocator.buffer(capacity: 5)
```

Comment:
I wonder if we can be even cleverer with this and re-use the same fixed buffer.

Reply:
Done in https://github.com/grpc/grpc-swift/pull/1529/commits/f3f17a2ce03f811bbcb88082217cb3b8fe97aeef